File: AEDS2/Verde/TP2/PesquisaBinaria.java
import java.util.Map;
import java.util.ArrayList;
import java.util.HashMap;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.PrintWriter;


public class PesquisaBinaria {
    private ArrayList<Jogador> vetorJogadores = new ArrayList<>();
    private int comparacoes = 0;

    // Método para inserir um objeto Jogador no vetor.
    public void inserirVetor(Jogador jogador) {
        vetorJogadores.add(jogador);
    }

    // Método para ordenar o vetor de acordo com o nome dos Jogadores.
    public void ordernarVetor() {
        int n = vetorJogadores.size() - 1;
        for(int i = n; i > 0; i--) {
            for(int j = 0; j < i; j++) {
                String nome1 = vetorJogadores.get(j).getNome();
                String nome2 = vetorJogadores.get(j+1).getNome();
                int comparacaoNomes = nome1.compareTo(nome2);
                if(comparacaoNomes > 0) {
                    // Swap
                    Jogador aux = vetorJogadores.get(j);
                    vetorJogadores.set(j, vetorJogadores.get(j+1));
                    vetorJogadores.set(j+1, aux);
                }
            }
        }
    }


    // Métodos Recursivos para realizar a Pesquisa Binaria no vetor.
    public boolean pesquisaBin(String nome) {
        return pesquisaBin(nome, 0, vetorJogadores.size() - 1);
    }

    public boolean pesquisaBin(String nome, int esq, int dir) {
        int meio = (esq + dir)/2;
        Jogador jogador = vetorJogadores.get(meio);
        comparacoes++;
        int comparacaoNomes = nome.compareTo(jogador.getNome());
        if(esq > dir) {
            return false;
        }
        else if(comparacaoNomes == 0) {
            return true;
        }
        else if(comparacaoNomes > 0) {
            return pesquisaBin(nome, meio + 1, dir);
        }
        else {
            return pesquisaBin(nome, esq, meio - 1);
        }
        
        
    }

    // Método para criar o arquivo de registro de log.
    public void registroDeLog(String matricula, long tempoExecucao) {
        try (FileWriter fw = new FileWriter(matricula + "_binaria.txt");
             PrintWriter pw = new PrintWriter(fw)) {
            pw.println(matricula + "\t" + tempoExecucao + "\t" + comparacoes);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // Método para comparar duas strings e verificar se são iguais.
    public static boolean equalStrings(String str_1, String str_2) {
        if(str_1.length() != str_2.length()) {
            return false;
        }

        for(int i = 0; i < str_1.length(); i++) {
            if(str_1.charAt(i) != str_2.charAt(i)) {
                return false;
            }
        }

        return true;
    }

    
    static class Jogador {
        private int id;
        private String nome;
        private int altura;
        private int peso;
        private String universidade;
        private int anoNascimento;
        private String cidadeNascimento;
        private String estadoNascimento;
    
        // Construtor padrão da classe Jogador
        public Jogador() {
            this.id = 0;
            this.nome = "";
            this.altura = 0;
            this.peso = 0;
            this.universidade = "";
            this.anoNascimento = 0;
            this.cidadeNascimento = "";
            this.estadoNascimento = "";
        }
    
        // Construtor com parametros para cada atributo.
        public Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento,
                String cidadeNascimento, String estadoNascimento) {
            this.id = id;
            this.nome = nome;
            this.altura = altura;
            this.peso = peso;
            this.universidade = universidade;
            this.anoNascimento = anoNascimento;
            this.cidadeNascimento = cidadeNascimento;
            this.estadoNascimento = estadoNascimento;
        }
    
        // Método para obter o ID do jogador
        public int getId() {
            return id;
        }
    
        // Método para definir o ID do jogador
        public void setId(int id) {
            this.id = id;
        }
    
        // Método para obter o nome do jogador
        public String getNome() {
            return nome;
        }
    
        // Método para definir o nome do jogador
        public void setNome(String nome) {
            this.nome = nome;
        }
    
        // Método para obter a altura do jogador
        public int getAltura() {
            return altura;
        }
    
        // Método para definir a altura do jogador
        public void setAltura(int altura) {
            this.altura = altura;
        }
    
        // Método para obter o peso do jogador
        public int getPeso() {
            return peso;
        }
    
        // Método para definir o peso do jogador
        public void setPeso(int peso) {
            this.peso = peso;
        }
    
        // Método para obter a universidade do jogador
        public String getUniversidade() {
            return universidade;
        }
    
        // Método para definir a universidade do jogador
        public void setUniversidade(String universidade) {
            this.universidade = universidade;
        }
    
        // Método para obter o ano de nascimento do jogador
        public int getAnoNasc() {
            return anoNascimento;
        }
    
        // Método para definir o ano de nascimento do jogador
        public void setAnoNasc(int anoNascimento) {
            this.anoNascimento = anoNascimento;
        }
    
        // Método para obter a cidade de nascimento do jogador
        public String getCidadeNasc() {
            return cidadeNascimento;
        }
    
        // Método para definir a cidade de nascimento do jogador
        public void setCidadeNasc(String cidadeNascimento) {
            this.cidadeNascimento = cidadeNascimento;
        }
    
        // Método para obter o estado de nascimento do jogador
        public String getEstadoNasc() {
            return estadoNascimento;
        }
    
        // Método para definir o estado de nascimento do jogador
        public void setEstadoNasc(String estadoNascimento) {
            this.estadoNascimento = estadoNascimento;
        }
    
        // Método para criar uma cópia (clone) do objeto Jogador
        public Jogador clone() {
            return new Jogador(id, nome, altura, peso, universidade, anoNascimento, cidadeNascimento, estadoNascimento);
        }
    
        // Método para ler dados de um arquivo CSV e preencher um mapa de jogadores
        public void ler(String nomeArquivo, Map<Integer, Jogador> map) {
            try (FileReader fileReader = new FileReader(nomeArquivo);
                    BufferedReader bufferedReader = new BufferedReader(fileReader)) {
                String linha;
                boolean cabecalho = true;
        
                while ((linha = bufferedReader.readLine()) != null) {
                    if (cabecalho) {
                        cabecalho = false;
                        continue; 
                    }
        
                    String[] parts = new String[8];
                    int contador = 0;
                    int inicioCampo = 0;
        
                    for (int i = 0; i < linha.length(); i++) {
                        if (linha.charAt(i) == ',') {
                            if (i == inicioCampo) {
                                parts[contador] = "nao informado"; 
                            } else {
                                parts[contador] = linha.substring(inicioCampo, i).trim();
                            }
                            contador++;
                            inicioCampo = i + 1;
                        }
                    }
                    
                    if (linha.charAt(linha.length() - 1) == ',') {
                        parts[contador] = "nao informado"; 
                    } else {
                        parts[contador] = linha.substring(inicioCampo).trim();
                    }
        
                    
                    if (contador != 7) {
                        continue;
                    }
        
                    int id = Integer.parseInt(parts[0]);
                    String nome = parts[1];
                    int altura = Integer.parseInt(parts[2]);
                    int peso = Integer.parseInt(parts[3]);
                    String universidade = parts[4];
                    int anoNascimento = Integer.parseInt(parts[5]);
                    String cidadeNascimento = parts[6];
                    String estadoNascimento = parts[7];
        
                    Jogador jogador = new Jogador(id, nome, altura, peso, universidade, anoNascimento, cidadeNascimento,
                            estadoNascimento);
                    map.put(id, jogador);
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

            
        
    
    }
    // Main
    public static void main(String[] args) {
        long tempoInicio = System.currentTimeMillis();
        PesquisaBinaria pesquisaBinaria = new PesquisaBinaria();
        Map<Integer, Jogador> jogadores = new HashMap<>();
        Jogador jogador = new Jogador();
        jogador.ler("/tmp/players.csv", jogadores);
        String entrada1, entrada2;
        do {
            entrada1 = MyIO.readLine();
            if(!equalStrings(entrada1, "FIM")) {
                int idInsercao = Integer.parseInt(entrada1);
                Jogador jogadorInsercao = jogadores.get(idInsercao);
                pesquisaBinaria.inserirVetor(jogadorInsercao);
            }
        }while(!equalStrings(entrada1, "FIM"));

        pesquisaBinaria.ordernarVetor();

         do {
            entrada2 = MyIO.readLine();
            if(!equalStrings(entrada2, "FIM")) {
                String nome = entrada2;
                boolean jogadorEncontrado = pesquisaBinaria.pesquisaBin(nome);
                if(jogadorEncontrado) {
                    MyIO.println("SIM");
                }
                else {
                    MyIO.println("NAO");
                }
                
            }
        }while(!equalStrings(entrada2, "FIM"));
        long tempoFinal = System.currentTimeMillis();
        long tempoExecucao = tempoFinal - tempoInicio;
        pesquisaBinaria.registroDeLog("800854", tempoExecucao);
        

    }
}

File: AEDS2/Verde/TP2/Q04/PesquisaBinaria.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

#define MAX 250

// Struct Jogador
typedef struct {
    int id;
    char nome[MAX];
    int altura;
    int peso;
    char universidade[MAX];
    int anoNascimento;
    char cidadeNascimento[MAX];
    char estadoNascimento[MAX];

}Jogador;

/* 
    Função para ler jogadores de um arquivo CSV e preencher uma lista de jogadores.
    O arquivo deve ter um cabeçalho e os campos separados por vírgulas.
*/
void ler(const char *nomeArquivo, Jogador** pJogadores, int* pNumJogadores) {
    int tamanhoAtual = 10;
    *pJogadores = (Jogador*) malloc(tamanhoAtual * sizeof(Jogador));
    FILE *arquivo = fopen(nomeArquivo, "r");
    if (arquivo == NULL) {
        perror("Erro ao abrir o arquivo");
        return;
    }

    char linha[1000];
    int cabecalho = 1;

    while (fgets(linha, sizeof(linha), arquivo) != NULL) {
        if (cabecalho) {
            cabecalho = 0;
            continue;
        }

        int inicioCampo = 0;
        int contador = 0;
        int k = 0;
        char* parts[8];
        

        for(int i = 0; i < strlen(linha); i++) {
            if(linha[i] == ',') {
                if(i == inicioCampo) {
                    parts[contador] = "nao informado";
                }
                else {
                    char aux[i - inicioCampo];
                    for(int j = inicioCampo; j < i; j++) {
                        aux[k] = linha[j];
                        k++;
                    }
                    aux[k] = '\0';
                    parts[contador] = strdup(aux);

                }
                k = 0;
                contador++;
                inicioCampo = i + 1;
                
            }
        }
        
        // Tratar o que sobra após a última vírgula como um campo separado
        if (inicioCampo < strlen(linha) - 1) {
            char aux[1000];
            int k = 0;
            for (int j = inicioCampo; j < strlen(linha); j++) {
                aux[k] = linha[j];
                k++;
            }
            aux[k] = '\0';
            if (k > 0 && aux[k - 1] == '\n') {
                aux[k - 1] = '\0';
            }
            parts[contador] = strdup(aux);
            contador++;
        }
        else {
            parts[contador] = "nao informado";
            contador++;
        }
        
        Jogador jogador;
        jogador.id = atoi(parts[0]);
        strcpy(jogador.nome, parts[1]);
        jogador.altura = atoi(parts[2]);
        jogador.peso = atoi(parts[3]);
        strcpy(jogador.universidade, parts[4]);
        jogador.anoNascimento = atoi(parts[5]);
        strcpy(jogador.cidadeNascimento, parts[6]);
        strcpy(jogador.estadoNascimento, parts[7]);

        if (*pNumJogadores == tamanhoAtual) {
        tamanhoAtual *= 2;
        *pJogadores = (Jogador*) realloc(*pJogadores, tamanhoAtual * sizeof(Jogador));
        }
        (*pJogadores)[*pNumJogadores] = jogador;
        (*pNumJogadores)++;

        

        
    }


    fclose(arquivo);
    
    
}

// Função para inserir um objeto Jogador em um vetor Jogador*.
void inserirVetor(Jogador jogador, Jogador** jogadores, int *contadorVetor) {
    if(*jogadores == NULL) {
        *jogadores = (Jogador*) malloc(sizeof(Jogador));
    }
    else {
        *jogadores = (Jogador*) realloc(*jogadores, (*contadorVetor + 1) * sizeof(Jogador));
    }

    (*jogadores)[*contadorVetor] = jogador;
    (*contadorVetor)++;
}

// Função para ordenar um vetor de jogadores.
void ordenarVetor(Jogador* jogadores, int numJogadores) {
    int n = numJogadores - 1;
    for(int i = n; i > 0; i--) {
        for(int j = 0; j < i; j++) {
            char* nome1 = jogadores[j].nome;
            char* nome2 = jogadores[j+1].nome;
            int comparacaoNomes = strcmp(nome1, nome2);
            if(comparacaoNomes > 0) {
                // Swap
                Jogador aux = jogadores[j];
                jogadores[j] = jogadores[j+1];
                jogadores[j+1] = aux;
            }
        }
    }
}

// Função recursiva para realizar a pesquisa binária no vetor de jogadores.
bool pesquisaBinaria(char nome[], int esq, int dir, Jogador* jogadores, int* comparacoes) {
    (*comparacoes)++;
    if(esq <= dir) {
        int meio = (esq + dir)/2;
        Jogador jogador = jogadores[meio];
        int comparacoesNome = strcmp(nome, jogador.nome);
        if(comparacoesNome == 0) {
            return true;
        }
        else if(comparacoesNome > 0) {
            return pesquisaBinaria(nome, meio + 1, dir, jogadores, comparacoes);
        }
        else {
            return pesquisaBinaria(nome, esq, meio - 1, jogadores, comparacoes);
        }
    }
    return false;
}

// Função para imprimir as informações de um jogador no formato especificado.
void imprimir(Jogador* jogadores, int i) {
    printf("[%d ## %s ## %d ## %d ## %d ## %s ## %s ## %s]\n", jogadores[i].id, jogadores[i].nome, 
    jogadores[i].altura, jogadores[i].peso, jogadores[i].anoNascimento, jogadores[i].universidade, jogadores[i].cidadeNascimento, jogadores[i].estadoNascimento);
}

// Main
int main() {
    Jogador* jogadores = NULL;
    Jogador* vetorJogadores = NULL;
    char* entrada1 = (char*)malloc(250 * sizeof(char));
    char* entrada2 = (char*)malloc(250*sizeof(char));
    int contadorVetor = 0;
    int numJogadores = 0;
    int comparacoes = 0;
    clock_t startTime, endTime;
    ler("/tmp/players.csv", &jogadores, &numJogadores);
    while(strcmp(entrada1, "FIM") != 0) {
        scanf(" %[^\n]", entrada1);
        if(strcmp(entrada1, "FIM") != 0) {
            int num = atoi(entrada1);
            Jogador jogador = jogadores[num];
            inserirVetor(jogador, &vetorJogadores, &contadorVetor);
        }
    }
    ordenarVetor(vetorJogadores, contadorVetor);

    startTime = clock();
    while(strcmp(entrada2, "FIM") != 0) {
        scanf(" %[^\n]", entrada2);
        if(strcmp(entrada2, "FIM") != 0) {
            bool jogadorEncontrado = pesquisaBinaria(entrada2, 0, contadorVetor - 1, vetorJogadores, &comparacoes);
            if(jogadorEncontrado) {
                printf("SIM\n");
            }
            else {
                printf("NAO\n");
            }
        }
    }

    endTime = clock();

    double totalTime = (double)(endTime - startTime);

    FILE *logFile = fopen("800854_binaria.txt", "w");
    if (logFile != NULL) {
        fprintf(logFile, "800854\t%lf\t%d", totalTime, comparacoes);
        fclose(logFile);
    }

    free(entrada1);
    free(entrada2);
    free(jogadores);
    free(vetorJogadores);
    return 0;
}

